### PR TITLE
backend/fix : Add-time-based-charge-in-ride-fare

### PR DESCRIPTION
**Pull request by vsaimanohar**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [X] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
 Add-time-based-charge-in-ride-fare

### Additional Changes

- [X] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [X] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [X] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs [hunk 1]
```diff
@@ -202,9 +210,11 @@ calculateFareParameters params = do
             params.distance - baseDistance
               & (\dist -> if dist > 0 then Just dist else Nothing)
           mbExtraKmFare = processFPProgressiveDetailsPerExtraKmFare perExtraKmRateSections <$> mbExtraDistance
+          timeBasedChargeInfo = calculateTimeBasedFare . (Money . (\avgSpeedOfVechile -> params.distance.getMeters `div` avgSpeedOfVechile) <$> (.getKilometers)) =<< averageSpeedOfVehicle
```
[the remark this answers is not in the document]
Comment (0utkarsh) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs:213]: Fare is not Re. 1 for each min, so you need to multiply it by the fare/min

File: Backend/dev/migrations/dynamic-offer-driver-app/0263-payment-nudge.sql [hunk 2]
```diff
@@ -1,42 +0,0 @@
-CREATE TABLE IF NOT EXISTS atlas_driver_offer_bpp.merchant_push_notification
```
Comment (0utkarsh) [inline Backend/dev/migrations/dynamic-offer-driver-app/0263-payment-nudge.sql:1]: Why is this file removed?

**Review by 0utkarsh — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vsaimanohar) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs:213]: > Also need to throw proper error when `avgSpeedOfVechile` is zero

avgSpeedOfVechile wont be zero its a const fare_policy value we would set it > 0

[the remark this answers is not in the document]
Comment (0utkarsh) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs:213]: We need to handle. If somebody sets it to zero, compilation will pass and we will start getting exceptions. Positive validation should be done

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs [hunk 3]
```diff
@@ -152,6 +160,7 @@ calculateFareParameters params = do
   logTagInfo "FareCalculator" $ "Initiating fare calculation for organization " +|| params.farePolicy.merchantId ||+ " and vehicle variant " +|| params.farePolicy.vehicleVariant ||+ ""
   let fp = params.farePolicy
   id <- generateGUID
+  unless (isJust (fp.averageSpeedOfVehicle) && fromJust (fp.averageSpeedOfVehicle) > 0) $ throwError AverageVehicleSpeedLessThanZero
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs:163]: Partial functions should be avoided even though you have an isJust check. Better to use case match

**Review by 0utkarsh — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs [hunk 4]
```diff
@@ -152,6 +161,9 @@ calculateFareParameters params = do
   logTagInfo "FareCalculator" $ "Initiating fare calculation for organization " +|| params.farePolicy.merchantId ||+ " and vehicle variant " +|| params.farePolicy.vehicleVariant ||+ ""
   let fp = params.farePolicy
   id <- generateGUID
+  case fp.averageSpeedOfVehicle of
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs:164]: Remove it from here.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs [hunk 5]
```diff
@@ -271,6 +285,17 @@ calculateFareParameters params = do
               cgst = Just . realToFrac $ baseFee * platformFeeInfo'.cgst,
               sgst = Just . realToFrac $ baseFee * platformFeeInfo'.sgst
             }
+    calculateTimeBasedFare :: Meters -> Maybe HighPrecMoney -> Maybe Minutes -> Kilometers -> Maybe Money
+    calculateTimeBasedFare distance timeBasedCharge rideDuration avgSpeedOfVehicle = do
+      let distancePerKilometer = realToFrac @_ @Double distance.getMeters / 1000
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs:290]: Rename variables correctly.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs [hunk 6]
```diff
@@ -271,6 +285,17 @@ calculateFareParameters params = do
               cgst = Just . realToFrac $ baseFee * platformFeeInfo'.cgst,
               sgst = Just . realToFrac $ baseFee * platformFeeInfo'.sgst
             }
+    calculateTimeBasedFare :: Meters -> Maybe HighPrecMoney -> Maybe Minutes -> Kilometers -> Maybe Money
+    calculateTimeBasedFare distance timeBasedCharge rideDuration avgSpeedOfVehicle = do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs:289]: have a check of avgSpeedOfVehicle.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs [hunk 7]
```diff
@@ -271,6 +285,17 @@ calculateFareParameters params = do
...
+      let distancePerKilometer = realToFrac @_ @Double distance.getMeters / 1000
+      let avgSpeedOfVehicle' = realToFrac @_ @Double avgSpeedOfVehicle.getKilometers
+      let timeBasedCharge' = realToFrac @_ @Double (fromMaybe 1 timeBasedCharge).getHighPrecMoney
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs:292]: Default value should be `0`

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs [hunk 8]
```diff
@@ -271,6 +285,17 @@ calculateFareParameters params = do
...
+      let timeBasedFareInMinutes :: Int = round $ (distancePerKilometer / avgSpeedOfVehicle') * 60
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs:293]: estimatedTimeTakeInMinutes.

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs [hunk 9]
```diff
@@ -271,6 +284,24 @@ calculateFareParameters params = do
...
+            SEDAN -> avgSpeedOfVehicle.sedan.getKilometers
+            SUV -> avgSpeedOfVehicle.suv.getKilometers
+            HATCHBACK -> avgSpeedOfVehicle.hatchback.getKilometers
+            AUTO_RICKSHAW -> avgSpeedOfVehicle.auto_rickshaw.getKilometers
+            TAXI -> avgSpeedOfVehicle.taxi.getKilometers
+            TAXI_PLUS -> avgSpeedOfVehicle.taxi_plus.getKilometers
+
+      let distanceInKilometer = realToFrac @_ @Double distance.getMeters / 1000
+      let perMinuteRideExtraTimeCharge' = realToFrac @_ @Double (fromMaybe 0 perMinuteRideExtraTimeCharge).getHighPrecMoney
+      let estimatedTimeTakeInMinutes :: Int = round $ (distanceInKilometer / avgSpeedOfVehicle') * 60
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs:299]: `avgSpeedOfVehicle'` can be zero. We are not checking this?

File: Backend/dev/migrations/dynamic-offer-driver-app/0265-add-averageVechileSpeed-to-transport-config.sql [hunk 10]
```diff
@@ -0,0 +1,5 @@
+ALTER TABLE atlas_driver_offer_bpp.avg_speed_of_vehicle
+ADD COLUMN JSON;
+
+UPDATE atlas_driver_offer_bpp.transporter_config
+SET avg_speed_of_vehicle = '{"seda": 30, "suv": 30, "hatchback": null,"auto_rickshaw": null, "taxi" : null, "taxi_plus" : null}'::json
```
Comment (0utkarsh) [inline Backend/dev/migrations/dynamic-offer-driver-app/0265-add-averageVechileSpeed-to-transport-config.sql:5]: sedan typo

**Review by 0utkarsh — CHANGES_REQUESTED** [review]

**Review by hkmangla — APPROVED** [review]

**Pull request merged**